op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,256 @@
+	CrossDerivedFromBlockRef(chainID eth.ChainID, derived eth.BlockID) (derivedFrom eth.BlockRef, err error)
+
+	LocalSafe(eth.ChainID) (types.DerivedBlockSealPair, error)
+	CrossSafe(eth.ChainID) (types.DerivedBlockSealPair, error)
+
+	RewindLocalSafe(eth.ChainID, types.BlockSeal) error
+	RewindCrossSafe(eth.ChainID, types.BlockSeal) error
+	RewindLogs(chainID eth.ChainID, newHead types.BlockSeal) error
+
+	FindSealedBlock(eth.ChainID, uint64) (types.BlockSeal, error)
+	Finalized(eth.ChainID) (types.BlockSeal, error)
+}
+
+// Rewinder is responsible for handling the rewinding of databases to the latest common ancestor between
+// the local databases and L2 node.
+type Rewinder struct {
+	log     log.Logger
+	emitter event.Emitter
+	l1Node  l1Node
+	db      rewinderDB
+}
+
+func New(log log.Logger, db rewinderDB, l1Node l1Node) *Rewinder {
+	return &Rewinder{
+		log:    log.New("component", "rewinder"),
+		db:     db,
+		l1Node: l1Node,
+	}
+}
+
+func (r *Rewinder) AttachEmitter(em event.Emitter) {
+	r.emitter = em
+}
+
+func (r *Rewinder) OnEvent(ev event.Event) bool {
+	switch x := ev.(type) {
+	case superevents.RewindL1Event:
+		r.handleRewindL1Event(x)
+		return true
+	case superevents.LocalDerivedEvent:
+		r.handleLocalDerivedEvent(x)
+		return true
+	default:
+		return false
+	}
+}
+
+// handleRewindL1Event iterates known chains and checks each one for a reorg
+// If a reorg is detected, it will rewind the chain to the latest common ancestor
+// between the local-safe head and the finalized head.
+func (r *Rewinder) handleRewindL1Event(ev superevents.RewindL1Event) {
+	for _, chainID := range r.db.DependencySet().Chains() {
+		if err := r.rewindL1ChainIfReorged(chainID, ev.IncomingBlock); err != nil {
+			r.log.Error("failed to rewind L1 data:", "chain", chainID, "err", err)
+		}
+	}
+}
+
+// handleLocalDerivedEvent checks if the newly derived block matches what we have in our unsafe DB
+// If it doesn't match, we need to rewind the logs DB to the common ancestor between
+// the LocalUnsafe head and the new LocalSafe block
+func (r *Rewinder) handleLocalDerivedEvent(ev superevents.LocalDerivedEvent) {
+	// Get the block at the derived height from our unsafe chain
+	newSafeHead := ev.Derived.Derived
+	unsafeVersion, err := r.db.FindSealedBlock(ev.ChainID, newSafeHead.Number)
+	if err != nil {
+		r.log.Error("failed to get unsafe block at derived height", "chain", ev.ChainID, "height", newSafeHead.Number, "err", err)
+		return
+	}
+
+	// If the block hashes match, our unsafe chain is still valid
+	if unsafeVersion.Hash == newSafeHead.Hash {
+		return
+	}
+
+	// Try rewinding the logs DB to the parent of the new safe head
+	// If it fails with a data conflict walk back through the chain
+	// until we find a common ancestor or reach the finalized block
+	finalized, err := r.db.Finalized(ev.ChainID)
+	if err != nil {
+		r.log.Error("failed to get finalized block", "chain", ev.ChainID, "err", err)
+		return
+	}
```

Comment:
I left another note elsewhere, but it could be an issue (mostly to tests) to prevent reorgs when finalization isn't set, as it might not be present early in supervisor's lifetime. I don't think we need to worry about it right now though, tests can just ensure finalization is set.